### PR TITLE
BAEL-4145 Circular Buffer

### PR DESCRIPTION
**Pull request by priyank-sriv**
http://jira.baeldung.com/browse/BAEL-4145

### CODE REVIEW

File: data-structures/src/main/java/com/baeldung/circularbuffer/CircularBuffer.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public boolean offer(E element) {
+
+        if (isNotFull()) {
+
+            int nextWriteSeq = writeSequence + 1;
+            data[nextWriteSeq % capacity] = element;
+
+            writeSequence += 1;
```
Comment (rpvilao) [inline data-structures/src/main/java/com/baeldung/circularbuffer/CircularBuffer.java:28]: writeSequence++? :)

File: data-structures/src/main/java/com/baeldung/circularbuffer/CircularBuffer.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        return false;
+    }
+
+    public E poll() {
+
+        if (isNotEmpty()) {
+
+            E nextValue = data[readSequence % capacity];
+            readSequence += 1;
```
Comment (rpvilao) [inline data-structures/src/main/java/com/baeldung/circularbuffer/CircularBuffer.java:40]: same

File: data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+public class ProducerConsumerLiveTest {
+
+    private final String[] shapes = { "Circle", "Triangle", "Rectangle", "Square", "Rhombus", "Trapezoid", "Pentagon", "Pentagram", "Hexagon", "Hexagram" };
+
+    @Test
+    public void givenACircularBuffer_whenInterleavingProducerConsumer_thenElementsMatch() throws InterruptedException {
+        CircularBuffer<String> buffer = new CircularBuffer<String>(shapes.length);
+        String[] consumedShapes = new String[shapes.length];
+
+        Thread producer = new Thread(new Producer(shapes, buffer));
```
Comment (rpvilao) [inline data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java:18]: let's use the concurrent api (Executors...)

File: data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            }
+        }
+    }
+
+    static class Consumer implements Runnable {
+
+        private CircularBuffer<String> buffer;
+        private String[] consumedShapes;
+
+        public Consumer(String[] consumedShapes, CircularBuffer<String> buffer) {
```
Comment (rpvilao) [inline data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java:58]: this consumer seems to be generic, maybe we get a better var name and not bound to the example

File: data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        assertArrayEquals(shapes, consumedShapes);
+    }
+
+    static class Producer implements Runnable {
+
+        private String[] producerShapes;
+        private CircularBuffer<String> buffer;
+
+        public Producer(String[] producerShapes, CircularBuffer<String> buffer) {
+            this.producerShapes = producerShapes;
```
Comment (rpvilao) [inline data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java:36]: same issue as "consumer"

File: data-structures/src/main/java/com/baeldung/circularbuffer/CircularBuffer.java (hunk 2 above)
Comment (priyank-sriv) [reply data-structures/src/main/java/com/baeldung/circularbuffer/CircularBuffer.java:40]: done

File: data-structures/src/main/java/com/baeldung/circularbuffer/CircularBuffer.java (hunk 1 above)
Comment (priyank-sriv) [reply data-structures/src/main/java/com/baeldung/circularbuffer/CircularBuffer.java:28]: done

File: data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java (hunk 3 above)
Comment (priyank-sriv) [reply data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java:18]: changed

File: data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java (hunk 4 above)
Comment (priyank-sriv) [reply data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java:58]: done

File: data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java (hunk 5 above)
Comment (priyank-sriv) [reply data-structures/src/test/java/com/baeldung/circularbuffer/ProducerConsumerLiveTest.java:36]: done

**Pull request merged**